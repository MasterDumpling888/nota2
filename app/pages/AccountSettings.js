import React, { useState, useEffect } from 'react';
import { View, Text, TextInput, Button, StyleSheet, Alert, TouchableWithoutFeedback, Keyboard, TouchableOpacity } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import PageBox from '../components/PageBox';
import { auth, db } from '../firebaseConfig';
import { updateProfile, deleteUser } from 'firebase/auth';
import { doc, getDoc, updateDoc, deleteDoc } from 'firebase/firestore';
import Icon from 'react-native-vector-icons/Ionicons';
import { getFontSize } from '../responsiveFont';

function AccountSettings({ navigation }) {
  const [username, setUsername] = useState('');
  const [newUsername, setNewUsername] = useState('');

  useEffect(() => {
    const loadUserData = async () => {
      const user = auth.currentUser;
      if (user) {
        const userDoc = await getDoc(doc(db, 'users', user.uid));
        if (userDoc.exists()) {
          setUsername(userDoc.data().username);
        }
      }
    };

    loadUserData();
  }, []);

  const handleUpdateUsername = async () => {
    const user = auth.currentUser;
    if (user && newUsername) {
      try {
        await updateProfile(user, { displayName: newUsername });
        await updateDoc(doc(db, 'users', user.uid), { username: newUsername });
        setUsername(newUsername);
        setNewUsername('');
        Alert.alert('Success', 'Username updated successfully');
      } catch (error) {
        console.error('Failed to update username', error);
        Alert.alert('Error', 'Failed to update username');
      }
    }
  };

  const handleSignOut = async () => {
    try {
      await auth.signOut();
      navigation.navigate('Login');
    } catch (error) {
      console.error('Failed to sign out', error);
      Alert.alert('Error', 'Failed to sign out');
    }
  };

  const handleDeleteAccount = async () => {
    const user = auth.currentUser;
    if (user) {
      try {
        await deleteUser(user);
        await deleteDoc(doc(db, 'users', user.uid));
        navigation.navigate('SignUp');
      } catch (error) {
        console.error('Failed to delete account', error);
        Alert.alert('Error', 'Failed to delete account');
      }
    }
  };

  return (
    <TouchableWithoutFeedback onPress={() => { Keyboard.dismiss(); handleUpdateUsername(); }}>
      <SafeAreaView style={{ flex: 1 }}>
        <PageBox title="Account Settings" onClose={() => navigation.goBack()}>
          <View style={styles.container}>
            <View style={styles.userContainer}>
              <Icon name="person-circle-outline" size={56} color="white" />
              <TextInput
                style={styles.input}
                value={newUsername || username}
                onChangeText={setNewUsername}
                onBlur={handleUpdateUsername}
              />
              <Icon name="pencil" size={30} color="white" />
            </View>
            <View style={styles.subContainer}>
              <Icon name="heart-outline" size={30} color="#33FD0A" />
              <Text style={styles.subText}>My Space</Text>
            </View>
            <View style={styles.sectionContainer}>
              <TouchableOpacity style={styles.sectionRow} onPress={() => navigation.navigate('Notes')}>
                <Icon name="document-text-outline" size={24} color="#33FD0A" />
                <Text style={styles.sectionText}>Notes</Text>
              </TouchableOpacity>
              <TouchableOpacity style={styles.sectionRow} onPress={() => navigation.navigate('Quizes')}>
                <Icon name="book-outline" size={24} color="#33FD0A" />
                <Text style={styles.sectionText}>Quizes</Text>
              </TouchableOpacity>
            </View>
            <View style={styles.subContainer}>
              <Icon name="cog" size={30} color="#33FD0A" />
              <Text style={styles.subText}>Settings</Text>
            </View>
            <View style={styles.sectionContainer}>
              <TouchableOpacity style={styles.sectionRow} onPress={handleSignOut}>
                <Icon name="log-out-outline" size={24} color="#33FD0A" />
                <Text style={styles.sectionText}>Sign Out</Text>
              </TouchableOpacity>
              <TouchableOpacity style={styles.sectionRow} onPress={handleDeleteAccount}>
                <Icon name="trash-outline" size={24} color="red" />
                <Text style={[styles.sectionText, { color: 'red' }]}>Delete Account</Text>
              </TouchableOpacity>
            </View>
          </View>
        </PageBox>
      </SafeAreaView>
    </TouchableWithoutFeedback>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: 16,
  },
  text: {
    fontSize: getFontSize(24),
    fontWeight: 'bold',
  },
  subText: {
    fontSize: getFontSize(16),
    color: 'white',
    fontFamily: 'Raleway-SemiBold',
  },
  input: {
    flex: 1,
    fontSize: getFontSize(24),
    color: 'white',
    fontFamily: 'Raleway-Bold',
  },
  userContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 16,
    gap: 10
  },
  subContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 10,
    gap: 10
  },
  sectionContainer: {
    borderWidth: 1,
    borderColor: '#33FD0A',
    borderRadius: 5,
    marginBottom: 16,
  },
  sectionRow: {
    padding: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#33FD0A',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 10
  },
  sectionText: {
    fontSize: getFontSize(14),
    color: 'white',
    fontFamily: 'Raleway-SemiBold',
  },
});

export default AccountSettings;
