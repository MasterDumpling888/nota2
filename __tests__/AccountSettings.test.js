import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import AccountSettings from '../app/pages/AccountSettings';

const mockNavigation = { navigate: jest.fn(), goBack: jest.fn() };

test('renders Account Settings screen', () => {
    const { getByText } = render(<AccountSettings navigation={mockNavigation} />);
    expect(getByText('Account Settings')).toBeTruthy(); // Checking title
});

test('renders Sign Out button', () => {
    const { getByText } = render(<AccountSettings navigation={mockNavigation} />);
    expect(getByText('Sign Out')).toBeTruthy(); // Checking Sign Out button
});

test('renders Delete Account button', () => {
    const { getByText } = render(<AccountSettings navigation={mockNavigation} />);
    expect(getByText('Delete Account')).toBeTruthy(); // Checking Delete Account button
});

test('AccountSettings snapshot test', () => {
    const { toJSON } = render(<AccountSettings navigation={mockNavigation} />);
    expect(toJSON()).toMatchSnapshot();
});