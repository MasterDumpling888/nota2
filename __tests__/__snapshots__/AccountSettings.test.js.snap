// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSettings snapshot test 1`] = `
<RNCSafeAreaView
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#181818",
        "flex": 1,
        "padding": 24,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
          }
        }
      >
        <Icon
          color="#33FD0A"
          name="chevron-back-outline"
          size={24}
          style={
            {
              "marginBottom": 15,
              "padding": 0,
            }
          }
        />
        <Text
          style={
            {
              "color": "#33FD0A",
              "fontFamily": "Raleway-SemiBold",
              "fontSize": 17,
              "marginBottom": 16,
            }
          }
        >
          Account Settings
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "padding": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 10,
            "marginBottom": 16,
          }
        }
      >
        <Icon
          color="white"
          name="person-circle-outline"
          size={56}
        />
        <TextInput
          onBlur={[Function]}
          onChangeText={[Function]}
          style={
            {
              "color": "white",
              "flex": 1,
              "fontFamily": "Raleway-Bold",
              "fontSize": 22.5,
            }
          }
          value=""
        />
        <Icon
          color="white"
          name="pencil"
          size={30}
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 10,
            "marginBottom": 10,
          }
        }
      >
        <Icon
          color="#33FD0A"
          name="heart-outline"
          size={30}
        />
        <Text
          style={
            {
              "color": "white",
              "fontFamily": "Raleway-SemiBold",
              "fontSize": 15,
            }
          }
        >
          My Space
        </Text>
      </View>
      <View
        style={
          {
            "borderColor": "#33FD0A",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#33FD0A",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "gap": 10,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Icon
            color="#33FD0A"
            name="document-text-outline"
            size={24}
          />
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 13,
              }
            }
          >
            Notes
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#33FD0A",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "gap": 10,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Icon
            color="#33FD0A"
            name="book-outline"
            size={24}
          />
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 13,
              }
            }
          >
            Quizes
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 10,
            "marginBottom": 10,
          }
        }
      >
        <Icon
          color="#33FD0A"
          name="cog"
          size={30}
        />
        <Text
          style={
            {
              "color": "white",
              "fontFamily": "Raleway-SemiBold",
              "fontSize": 15,
            }
          }
        >
          Settings
        </Text>
      </View>
      <View
        style={
          {
            "borderColor": "#33FD0A",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#33FD0A",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "gap": 10,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Icon
            color="#33FD0A"
            name="log-out-outline"
            size={24}
          />
          <Text
            style={
              {
                "color": "white",
                "fontFamily": "Raleway-SemiBold",
                "fontSize": 13,
              }
            }
          >
            Sign Out
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#33FD0A",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "gap": 10,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Icon
            color="red"
            name="trash-outline"
            size={24}
          />
          <Text
            style={
              [
                {
                  "color": "white",
                  "fontFamily": "Raleway-SemiBold",
                  "fontSize": 13,
                },
                {
                  "color": "red",
                },
              ]
            }
          >
            Delete Account
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
